### PR TITLE
Enable heat transfer with solidification and two-phase flow

### PR DESCRIPTION
**Pull request by nmuch**
This PR enables heat transfer with solidification and two-phase flow. 

The liquid's material are determined with the level-set function. On top of that, the solidification material parameters are determined consistently. I.e. if the temperature is below the liquidus temperature the liquid's parameters are interpolated with the solid's material parameters. If the temperature is below the solidus temperature, the material parameters equal the solid's material parameters regardless of the level-set.

In the heat_transfer_operator the material parameters can be determined with the new `get_material_parameters()`. The tangent cell loop required the temperature derivatives of the material parameters too, so there is `get_material_parameters_and_derivatives()` that determines the derivatives in an efficient manner. 

Since density and capacity are only used in as a product in the heat equation, they are now combined to `rho_cp`.

The new test case `unidirectional_heat_transfer_with_two_phase_and_solidification` is a simple 1D simulation with a prescribed velocity field and a prescribed level-set filed that is transported with that velocity.

### CODE REVIEW

**Comment by nmuch** [comment]
I get some weird behavior: The new test `unidirectional_heat_transfer_with_two_phase_and_solidification` runs perfectly without mpirun or if `"paraview do output"` is set to `"true"`. 

However if I run is with mpirun and `"paraview do output": "false"` the linear solver won't converge. Does anyone have an idea why?

**Comment by peterrum** [comment]
> However if I run is with mpirun and "paraview do output": "false" the linear solver won't converge. Does anyone have an idea why?

Is it possible that you go into the postprocessor with a ghosted vector, the postprocessor clears the ghost values, and after that the ghost values are not updated again? At least this could explain the observed behavior.

File: source/heat/heat_transfer_operator.cpp [hunk 1]
```diff
@@ -841,23 +810,189 @@ namespace MeltPoolDG::Heat
...
+        get_material_parameters_with_solidification(capacity,
+                                                    conductivity,
+                                                    density,
+                                                    liq_capacity,
+                                                    liq_conductivity,
+                                                    liq_density,
+                                                    solid_fraction);
+        rho_cp = density * capacity;
+        return;
+      }
```
Comment (mschreter) [inline source/heat/heat_transfer_operator.cpp:878]: Could you simplify it a bit, e.g. as follows:

``` cpp
    if (!with_solidification && !with_two_phase)
      {
        rho_cp       = material.first.density * material.first.capacity;
        conductivity = material.first.conductivity;
        return;
      }

    VectorizedArray<number> liquid_capacity(material.second.capacity);
    VectorizedArray<number> liquid_density(material.second.density);
    VectorizedArray<number> liquid_conductivity(material.second.conductivity);
    
    if (with_two_phase)
    {
        get_material_parameters_with_two_phase_flow(liquid_capacity,
                                                    liquid_conductivity,
                                                    liquid_density,
                                                    ls_heaviside_val.get_value(q_index));
        rho_cp = liquid_density * liquid_capacity;
    }

    if (with_solidification)
      {
        VectorizedArray<number> capacity;
        VectorizedArray<number> density;
        const auto solid_fraction = calculate_solid_fraction(temp_lin_val.get_value(q_index));
        get_material_parameters_with_solidification(
          capacity,
          conductivity,
          density,
          liquid_capacity,
          liquid_conductivity,
          liquid_density,
          solid_fraction);
        rho_cp = density * capacity;
      }
```

File: source/heat/heat_transfer_operator.cpp [hunk 2]
```diff
@@ -841,23 +810,189 @@ namespace MeltPoolDG::Heat
...
+                                                               d_density_dT,
+                                                               liq_capacity,
+                                                               liq_conductivity,
+                                                               liq_density,
+                                                               solid_fraction);
+        d_rho_cp_dT = d_capacity_dT * density + d_density_dT * capacity;
+        return;
+      }
+  }
+
```
Comment (mschreter) [inline source/heat/heat_transfer_operator.cpp:976]: Maybe you could simplifiy this function in a similar fashion to the previous one.

File: include/meltpooldg/utilities/utilityfunctions.hpp [hunk 3]
```diff
@@ -38,6 +38,17 @@ namespace MeltPoolDG
...
+    MPI_Comm
+    get_mpi_comm(const MeshType &mesh)
+    {
+      const auto *tria_parallel =
+        dynamic_cast<const parallel::TriangulationBase<MeshType::dimension> *>(
+          &(mesh.get_triangulation()));
+
+      return tria_parallel != nullptr ? tria_parallel->get_communicator() : MPI_COMM_SELF;
+    }
+
```
[the remark this answers is not in the document]
Comment (nmuch) [reply include/meltpooldg/utilities/utilityfunctions.hpp]: Yes, I checked out an old version. I deleted it already.

File: include/meltpooldg/utilities/utilityfunctions.hpp [hunk 4]
```diff
@@ -326,13 +337,27 @@ namespace MeltPoolDG
       return val1 + (val2 - val1) * (-2. * ls * ls * ls + 3. * ls * ls);
     }
 
+    template <typename value_type>
+    inline value_type
+    interpolate_cubic(const value_type &ls, const value_type &val1, const double val2)
+    {
+      return val1 + (val2 - val1) * (-2. * ls * ls * ls + 3. * ls * ls);
+    }
+
```
[the remark this answers is not in the document]
Comment (nmuch) [reply include/meltpooldg/utilities/utilityfunctions.hpp:335]: Yes. I need a version that takes the `val1` and `val2` of type `double` and `double` and one version that takes `VectorizedArray` and `double`. Sadly a double cannot be automatically converted into a `VectorizedArray`.

File: include/meltpooldg/utilities/utilityfunctions.hpp [hunk 5]
```diff
@@ -326,13 +337,27 @@ namespace MeltPoolDG
...
       return (val2 - val1) * (-6. * ls * ls + 6. * ls);
     }
 
+    template <typename value_type>
+    inline value_type
+    interpolate_cubic_derivative(const value_type &ls, const value_type &val1, const double val2)
+    {
+      return (val2 - val1) * (-6. * ls * ls + 6. * ls);
+    }
+
```
[the remark this answers is not in the document]
Comment (nmuch) [reply include/meltpooldg/utilities/utilityfunctions.hpp:349]: Same here.

File: source/heat/heat_transfer_operator.cpp [hunk 6]
```diff
@@ -439,11 +436,8 @@ namespace MeltPoolDG::Heat
     FECellIntegrator<dim, 1, number>   ls_vals(matrix_free, ls_dof_idx, this->quad_idx);
     FECellIntegrator<dim, 1, number>   evapor_vals(matrix_free, temp_dof_idx, this->quad_idx);
 
-    VectorizedArray<number> capacity     = material.first.capacity;
-    VectorizedArray<number> conductivity = material.first.conductivity;
-    VectorizedArray<number> density      = material.first.density;
-
     VectorizedArray<number> rho_cp;
```
[the remark this answers is not in the document]
Comment (nmuch) [reply source/heat/heat_transfer_operator.cpp:439]: The value of `rho_cp` and `conductivity` will always be filled by `get_material_parameters()` so we don't need to initialize a value.

File: source/heat/heat_transfer_operator.cpp [hunk 7]
```diff
@@ -687,16 +672,10 @@ namespace MeltPoolDG::Heat
...
-    VectorizedArray<number> density      = material.first.density;
-
-    VectorizedArray<number> d_capacity_dT     = 0.0;
-    VectorizedArray<number> d_conductivity_dT = 0.0;
-    VectorizedArray<number> d_density_dT      = 0.0;
-
     VectorizedArray<number> rho_cp;
+    VectorizedArray<number> conductivity;
     VectorizedArray<number> d_rho_cp_dT;
+    VectorizedArray<number> d_conductivity_dT;
```
[the remark this answers is not in the document]
Comment (nmuch) [reply source/heat/heat_transfer_operator.cpp:678]: No, see above ;)

**Comment by mschreter** [comment]
> I get some weird behavior: The new test `unidirectional_heat_transfer_with_two_phase_and_solidification` runs perfectly without mpirun or if `"paraview do output"` is set to `"true"`.
> 
> However if I run is with mpirun and `"paraview do output": "false"` the linear solver won't converge. Does anyone have an idea why?

It seems like a parallel bug (independent of paraview do output). With 1 process I don't have any problem. However, using n+1 processes produces increasingly bad convergence of the linear solver.

File: source/heat/heat_transfer_operator.cpp (hunk 1 above)
Comment (nmuch) [reply source/heat/heat_transfer_operator.cpp:878]: First I separated all cases to see what combinations of `with_two_phase` and `with_solidification` look like. We can probably improve performatnve with your version. If ´!with_solidification && !with_two_phase` the this function can directly return if the values are initialized correctly.

File: include/meltpooldg/utilities/utilityfunctions.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (mschreter) [reply include/meltpooldg/utilities/utilityfunctions.hpp:335]: Mmh. You could also use additional typenames (although it's not elegant, however at least saves duplicate code):

```suggestion
    template <typename value_type1, typename value_type2, typename value_type3>
    inline value_type
    interpolate_cubic(const value_type1 &ls, const value_type2 &val1, const value_type3 val2)
    {
      return val1 + (val2 - val1) * (-2. * ls * ls * ls + 3. * ls * ls);
    }

```

**Comment by nmuch** [comment]
> It seems like a parallel bug (independent of paraview do output). With 1 process I don't have any problem. However, using n+1 processes produces increasingly bad convergence of the linear solver.

I spend some time trying to fix this issue but I can't find it. Commenting out `attach_output_vectors()` did nothing, so did adding `update_ghost_values()` before ` heat_operation->solve(dt);`.

I figure this problem has nothing to do with this PR, because no changes effect the postprocessor or linear solver.

A quick workaround is to change the test so it's run on one core only. We can fix the issue in a followup.

**Pull request merged**